### DISCUSSION TITLE
📢 CSLA 5.4.0 preview is online (final preview?)

### DISCUSSION THREAD

**Discussion by rockfordlhotka** [MarimerLLC/csla#1898] [Announcements]
I just pushed a preview of CSLA 5.4.0 to NuGet.

This preview includes support for:

* .NET 5
* .NET 5 Blazor
* .NET 5 ASP.NET
* .NET 5 WPF (`Csla.Xaml`)
* .NET 5 Windows Forms

Because Microsoft no longer supports netcoreapp3.0, the project targeting 3.0 now target 3.1:

* aspnetcore
* WPF (`Csla.Xaml`)
* Windows Forms

The [release notes](https://github.com/MarimerLLC/csla/blob/master/releasenotes.md) provides more detail on other changes.

**Comment by BlagoCuljak** [comment]
Hi Rocky, I'm having issue with testing preview, since the Csla.AspNetCore is not in preview, and it requires Csla (= 5.3.2).

Even if I install Csla preview manually, Csla.AspNetCore still doesn't play with 5.4 preview. 

Any chance of bumping the Csla.AspNetCore to preview and using the Csla preview?

**Reply by rockfordlhotka** [reply]
Hmm. I did update Csla.AspNetCore to have a target for `net5.0` as well as the pre-existing `netcoreapp3.1`. So it _should_ be the case that when you reference the package from a `net5.0` web project it'll use the correct csla assembly.

Can you look at your csproj file and confirm the `TargetFramework` element for me?

**Reply by BlagoCuljak** [reply]
I don't see preview [here](https://www.nuget.org/packages/Csla.AspNetCore/).

**Reply by rockfordlhotka** [reply]
Thank you!

Should be there now.

**Reply by BlagoCuljak** [reply]
Hi Rocky,
I tried for 3 hours to get the compression working, but I don't see my changes. I don't see headers in response or request.

I have reversed Csla.dll, and Csla.AspNetCore, both of them 5.4 version, and both of them lack headers. Can you please check your Release pipeline to NuGet?

![server](https://user-images.githubusercontent.com/34435929/98930400-e054f000-24dc-11eb-9921-04dd48b716fb.jpg)
![client](https://user-images.githubusercontent.com/34435929/98930405-e21eb380-24dc-11eb-8833-b24400a249b0.jpg)

**Reply by rockfordlhotka** [reply]
Sorry @BlagoCuljak! I appreciate your persistence!

I'm pushing a new prerelease that _should_ include the latest code changes. I suspect you are right, and that something went wrong in my build pipeline. 🤞

**Reply by BlagoCuljak** [reply]
Hi Rocky, great news! Now I see changes, and it works as it should.

Even if the Startup.cs hasn't got enabled compression, it continues to work.

However, one bad news, Webassembly is not working. It's stuck on loading, and I don't see any requests. Disabling compression has the same effect, and I don't think it's compression/headers to blame.

**Reply by rockfordlhotka** [reply]
Glad to hear compression is working!

With wasm, are you using .NET 5 or 3.1?

**Reply by BlagoCuljak** [reply]
3.1

**Comment by rockfordlhotka** [comment]
📢 I just updated the preview to 20113004, which hopefully is the final preview.

**Please test if you can** to help ensure this preview works for your scenarios.

**Reply by BlagoCuljak** [reply]
Just tested new version, everything works well!

**Reply by JacoJordaan** [reply]
> 📢 I just updated the preview to 20113004, which hopefully is the final preview.
> 
> **Please test if you can** to help ensure this preview works for your scenarios.

Tested from my side, but this is not working.  Version 5.3.4 was working fine.
Get the following error on retrieval of business object:
Message: ````DataPortal.Fetch failed on the server````
Stack trace: 
````
      at Csla.DataPortalClient.HttpProxy.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.DataPortal`1.DoFetchAsync(Type objectType, Object criteria, Boolean isSync)
   at Csla.DataPortal`1.FetchAsync(Object[] criteria)
   at Csla.DataPortal.FetchAsync[T]()
````
![image](https://user-images.githubusercontent.com/43467554/100713720-64ecad00-33bd-11eb-982d-95fc246fc733.png)

I do use my own implementation of IDataPortalProxy. This has never been an issue. Not sure if something has changed??
````
public class SmartHttpProxy : IDataPortalProxy
    {
        private Csla.DataPortalClient.HttpProxy GetProxy(Type objectType)
        {
            string serverName = GetServerName(objectType);m

            string serverUrl = GetAppServerUrl(serverName);

            return new Csla.DataPortalClient.HttpProxy(serverUrl);
        }

        internal static string GetServerName(Type objectType)
        {
            ...
        }

        internal static string GetAppServerUrl(Type objectType)
        {
            ...
        }
        

        public bool IsServerRemote
        {
            get { return true; }
        }

        public Task<DataPortalResult> Create(Type objectType, object criteria, DataPortalContext context, bool isSync)
        {
            return GetProxy(objectType).Create(objectType, criteria, context, isSync);
        }

        public Task<DataPortalResult> Fetch(Type objectType, object criteria, DataPortalContext context, bool isSync)
        {
            return GetProxy(objectType).Fetch(objectType, criteria, context, isSync);
        }

        public Task<DataPortalResult> Update(object obj, DataPortalContext context, bool isSync)
        {
            return GetProxy(obj.GetType()).Update(obj, context, isSync);
        }

        public Task<DataPortalResult> Delete(Type objectType, object criteria, DataPortalContext context, bool isSync)
        {
            return GetProxy(objectType).Delete(objectType, criteria, context, isSync);
        }
    }
````

**Reply by rockfordlhotka** [reply]
Can you catch the exception and get its ToString value? That'll give us the full (server and client) stack trace.

Or since you seem to have a break point on the exception, you can use the Immediate window to do a ToString.

**Reply by JacoJordaan** [reply]
Looks like `NullReferenceException: service` is the issue.
````
Csla.DataPortalException: DataPortal.Fetch failed on the server

   at Csla.DataPortalClient.HttpProxy.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.DataPortal`1.DoFetchAsync(Type objectType, Object criteria, Boolean isSync)
   at Csla.DataPortal`1.FetchAsync(Object[] criteria)
   at Csla.DataPortal.FetchAsync[T]()
   at Symbioses.QS.Business.Shared.CompanyInfo.GetInfoAsync() in C:\Users\Symbioses Technology\source\repos\QS\Symbioses.QS.Business.Shared\BO\Quote\CompanyInfo.cs:line 213
------------------------------
Csla.Server.DataPortalException: DataPortal.Fetch failed on the server
   at Csla.Reflection.ServiceProviderMethodCaller.CallMethodTryAsync(Object obj, ServiceProviderMethodInfo method, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
   at Csla.Server.DataPortalTarget.InvokeOperationAsync[T](Object criteria, Boolean isSync)
   at Csla.Server.DataPortalTarget.FetchAsync(Object criteria, Boolean isSync)
   at Csla.Server.SimpleDataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.Server.DataPortal.NewDataPortalException(String message, Exception innerException, Object businessObject)
   at Csla.Server.SimpleDataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.Server.DataPortalSelector.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.Server.DataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.Server.Hosts.HttpPortal.Fetch(CriteriaRequest request)
   at Csla.Server.Hosts.HttpPortalController.CallPortal(String operation, Object request)
   at Csla.Server.Hosts.HttpPortalController.InvokePortal(String operation, Stream requestStream, Stream responseStream)
Csla.Reflection.CallMethodException: CompanyInfo.DataPortal_Fetch method call failed
   at Csla.Reflection.ServiceProviderMethodCaller.CallMethodTryAsync(Object obj, ServiceProviderMethodInfo method, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
   at Csla.Server.DataPortalTarget.InvokeOperationAsync[T](Object criteria, Boolean isSync)
   at Csla.Server.DataPortalTarget.FetchAsync(Object criteria, Boolean isSync)
   at Csla.Server.SimpleDataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
System.NullReferenceException: service
   at Csla.Reflection.ServiceProviderMethodCaller.CallMethodTryAsync(Object obj, ServiceProviderMethodInfo method, Object[] parameters)
   at Csla.Reflection.LateBoundObject.CallMethodTryAsyncDI[T](Boolean isSync, Object[] parameters)
````

**Reply by rockfordlhotka** [reply]
Ahh, thank you! The problem is related to #1874. Hopefully @dazinator has an idea what's going on, otherwise it'll need some research.

Just to confirm, @JacoJordaan, your app server startup code does call `AddCsla` and `UseCsla` right?

**Reply by dazinator** [reply]
I have a sneaking suspicion that this is because @JacoJordaan is using a remote data portal, and his business objects probably use the new `inject` attribute to have dependencies injected by the server side data portal, but the remote application host probably isn't creating a DI scope for the incoming CSLA data portal request snd setting that as the active IServiceProvider for csla to use - if csla is not aware of an IServiceProvider then it will throw a null ref error at the point its about to try and inject dependencies. @rockfordlhotka I remember us talking about the various hosts needing to be adjusted to make sure a DI scope / service provider was available for the data portal request / response so that injection worked on the host, but I dont know if that actually got done. If your remote host is an aspnet core host, and the data portal can run as an mvc controller, then aspnet core already creates this scope, and csla can resolve it via IHttpContextAccessor so that should just all work.. but if the host is some other application- you'll need to create a new service provider scope for the inbound data portal request, call a method that sets it as the current scoped IServiceProvider for csla, then dispose of that scope once the dataportal response has been returned to the client - if you want to use injection on the remote dp host.

**Reply by rockfordlhotka** [reply]
I remember that conversation @dazinator, but somehow didn't make the connection that this was a breaking change for existing code.

Exactly what code does @JacoJordaan need to put in his `Startup.cs` to make this all work?

(and me in ProjectTracker, which will break with 5.4.0 too I am sure)

**Reply by JacoJordaan** [reply]
Just to confirm, I am using the remote data portal with the new `inject` attribute in the business objects server side calls. 
Startup code of application server calls `AddCsla` and 'UseCsla`.
Basically the same as the ProjectTracker samples.

**Reply by rockfordlhotka** [reply]
Thank you @JacoJordaan - I think you may need to add a line or two to your startup.cs on the appserver project to capture the auto-generated `ScopedServiceProvider`. Hopefully @dazinator can help refresh my memory on how that's done.

**Reply by dazinator** [reply]
> Exactly what code does @JacoJordaan need to put in his Startup.cs to make this all work?

If he is hosting the DataPortal as an MVC controller from an aspnet core host, then nothing - CSLA should detect current request scope for the dp request, via IHttpContextAccessor.

@JacoJordaan is your server side data portal being hosted as an MVC controller? Here is an example from the simplentier sample:

```
[Route("api/[controller]")]
  [ApiController]
  public class DataPortalController : Csla.Server.Hosts.HttpPortalController
  {
    [HttpGet]
    public string Get()
    {
      return "Running...";
    }
  }

```

and 

```
 public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
      if (env.IsDevelopment())
      {
        app.UseDeveloperExceptionPage();
      }

      app.UseHttpsRedirection();

      app.UseRouting();

      app.UseAuthorization();

      app.UseEndpoints(endpoints =>
      {
        endpoints.MapControllers();
      });
    }
```

**Comment by dazinator** [comment]
@rockfordlhotka 
As per above, if the data portal can be hosted and routed to via asp.net core endpoint routing, then it will be running within the asp.net core middleware pipeline, and as such, `IHttpContextAccessor` will be used to get the current request scoped IServiceProvider and CSLA injection will work - this code is already in place.

The issue will be, where the DataPortal is not running within an asp.net core endpoint routing / middleware pipeline. For those hosts, and only if you want to use [inject] - code must be added to the processing of the request at a suitable place - exact code will be different based on tech used to process the request (WCF, vs OWIN, vs Nancy, etc etc)

```
var requestHandler => (inboundHttpRequest, innerRequestHandler)=>{ 
 // create or get the IServiceProvider scoped for current request
using(var scope = applicationlevelServiceProvoder.CreateScope())
{
   ApplicationContext.ServiceProviderScope = scope.ServiceProvider; // this scoped IServiceProvider csla will use to inject dependencies.
   innerRequestHandler(inboundHttpRequest); // request should be handled downstream by data portal.
} // scope is disposed at end of request. 
}
```

**Reply by dazinator** [reply]
@JacoJordaan if you are still hitting issues, it would help if you could share see exactly how your server side data portal is defined & executed - is it an mvc controller? is it hosted via wcf? etc etc - as it doesn't look like there are any test cases in place to verify this scenario - perhaps a minimal repro if you have time.

**Reply by JacoJordaan** [reply]
I found the issue. I have multiple application servers. With the one I did not use `app.UseCsla()`. 
Clearly required for version 5.4.0.
Sincere apologies for the misdirect.

**Reply by rockfordlhotka** [reply]
That's great to hear!! I am feeling a lot of relief!! 😌

**Comment by rockfordlhotka** [comment]
Hello everyone!

At this time I plan to go live with 5.4.0 next Tuesday US central time. 

If you have time to do any testing of 5.4.0 in your environments and apps, please do so and let me know the results.

Thank you!

**Reply by rockfordlhotka** [reply]
Still planning to go live today - probably in about 2-3 hours depending on how my day unfolds.

**Comment by uwefms** [comment]
Hi everybody,

I was able to convert the Blazor part of ProjectTracker to net5.0. with 5.4.0. The issue I was reporting (#1911) could be resolved by changing 

`      <NotAuthorized>
		<a href="/Account/Login">Log in</a>
      </NotAuthorized>
`
to

`      <NotAuthorized>
        @if (App.IsServerSide)
        {
		  <form action="/Account/Login">            
			<button type="submit" class="nav-link btn btn-link">Log in</button>
          </form>
        }
        else
        {
		<a href="/Account/Login">Log in</a>
	}
      </NotAuthorized>
`
in MainLayout.razor. 

Everything seems to be working fine. 

The only thing I am usure about is the change from

`builder.RootComponents.Add<App>("app") to builder.RootComponents.Add<App>("#app")`

in (WASM) Program.Main

and

`<app>Loading...</app> to <div id="app">Loading...</div>`

in index.html which was not working because app.razor is in another project. 
Could somebody explain to me why (or if) this change is necessary ?

  regards

     Uwe

**Reply by rockfordlhotka** [reply]
I haven't completely figured out the `#app` thing either.

The new .NET 5 templates all use that, and it works. But when I've tried to manually update older projects, I can't figure out where the root component ties into the rest of the app - in other words, I don't know where else you need to change `app` to `#app` to make it work.

So what I've been doing for all the projects so far is to copy the existing project, delete it, create new starter projects using the .NET 5 templates, then copy back in the bits of original code/pages/etc. to the new/blank project.

**Reply by uwefms** [reply]
Hi Rocky,
 thanks for the clarification. Will do the same and keep testing. 

  regards

      Uwe

**Reply by mtavares628** [reply]
Just an FYI for this, the change needed for #app is found in wwwroot/index.html of the client project.  
Change
`<app>Loading...</app>`
To:
`<div id="app">Loading...</div>`

I also had to perform a ctrl F5 as the page was still cached with the old markup.  The full steps for migrating to .NET 5 are here:

(https://docs.microsoft.com/en-us/aspnet/core/migration/31-to-50?view=aspnetcore-5.0&tabs=visual-studio#update-blazor-webassembly-projects)

Incidentally, I'm having trouble with the Dataportal and DI after updating to 5.4.0, but I'll start a new thread for that as this thread references the preview.
